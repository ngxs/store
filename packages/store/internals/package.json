{
  "ngPackage": {
    "$schema": "../../../node_modules/ng-packagr/ng-package.schema.json",
    "lib": {
      "entryFile": "src/index.ts",
      "flatModuleFile": "ngxs-store-internals",
      "umdModuleIds": {}
    }
  }
}
