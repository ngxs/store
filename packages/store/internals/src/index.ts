export { isAngularInTestMode } from './angular';
export { NgxsBootstrapper } from './ngxs-bootstrapper';
export { memoize } from './memoize';
export { INITIAL_STATE_TOKEN, InitialState } from './initial-state';
export { PlainObjectOf, PlainObject, StateClass } from './symbols';
export { NGXS_STATE_CONTEXT_FACTORY, NGXS_STATE_FACTORY } from './internal-tokens';
