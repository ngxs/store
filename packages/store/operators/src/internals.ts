export type Predicate<T = any> = (value?: T) => boolean;
