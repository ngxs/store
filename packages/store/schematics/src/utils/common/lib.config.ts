export enum LIBRARIES {
  DEVTOOLS = '@ngxs/devtools-plugin',
  FORM = '@ngxs/form-plugin',
  HMR = '@ngxs/hmr-plugin',
  LOGGER = '@ngxs/logger-plugin',
  ROUTER = '@ngxs/router-plugin',
  STORAGE = '@ngxs/storage-plugin',
  STORE = '@ngxs/store',
  WEBSOCKET = '@ngxs/websocket-plugin'
}
