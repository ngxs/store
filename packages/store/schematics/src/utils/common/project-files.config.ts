export const PACKAGE_JSON = 'package.json';
export const ANGULAR_JSON = 'angular.json';
export const TSCONFIG_SPEC_JSON = 'src/tsconfig.spec.json';
