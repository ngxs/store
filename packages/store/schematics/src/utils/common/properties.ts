export function isEmpty(value: string | undefined): boolean {
  return typeof value !== 'undefined' && !value.trim().length;
}
