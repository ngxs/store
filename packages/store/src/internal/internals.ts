import { META_KEY, ActionOptions, SELECTOR_META_KEY } from '../symbols';
import { Observable } from 'rxjs';

export interface ObjectKeyMap<T> {
  [key: string]: T;
}

export interface StateClass {
  [META_KEY]?: MetaDataModel;
}

export type StateKeyGraph = ObjectKeyMap<string[]>;

export interface ActionHandlerMetaData {
  fn: string;
  options: ActionOptions;
  type: string;
}

export interface StateOperations<T> {
  getState(): T;
  setState(val: T);
  dispatch(actions: any | any[]): Observable<void>;
}

export interface MetaDataModel {
  name: string;
  actions: ObjectKeyMap<ActionHandlerMetaData[]>;
  defaults: any;
  path: string;
  children: StateClass[];
  instance: any;
}

export type SelectFromState = (state: any) => any;

export interface SelectorMetaDataModel {
  selectFromAppState: SelectFromState;
  originalFn: Function;
  containerClass: any;
  selectorName: string;
}

export interface MappedStore {
  name: string;
  actions: ObjectKeyMap<ActionHandlerMetaData[]>;
  defaults: any;
  instance: any;
  depth: string;
}

/**
 * Ensures metadata is attached to the class and returns it.
 *
 * @ignore
 */
export function ensureStoreMetadata(target): MetaDataModel {
  if (!target.hasOwnProperty(META_KEY)) {
    const defaultMetadata: MetaDataModel = {
      name: null,
      actions: {},
      defaults: {},
      path: null,
      children: [],
      instance: null
    };

    Object.defineProperty(target, META_KEY, { value: defaultMetadata });
  }
  return getStoreMetadata(target);
}

/**
 * Get the metadata attached to the class if it exists.
 *
 * @ignore
 */
export function getStoreMetadata(target): MetaDataModel {
  return target[META_KEY];
}

/**
 * Ensures metadata is attached to the selector and returns it.
 *
 * @ignore
 */
export function ensureSelectorMetadata(target): SelectorMetaDataModel {
  if (!target.hasOwnProperty(SELECTOR_META_KEY)) {
    const defaultMetadata: SelectorMetaDataModel = {
      selectFromAppState: null,
      originalFn: null,
      containerClass: null,
      selectorName: null
    };

    Object.defineProperty(target, SELECTOR_META_KEY, { value: defaultMetadata });
  }

  return getSelectorMetadata(target);
}

/**
 * Get the metadata attached to the selector if it exists.
 *
 * @ignore
 */
export function getSelectorMetadata(target): SelectorMetaDataModel {
  return target[SELECTOR_META_KEY];
}

/**
 * The generated function is faster than:
 * - pluck (Observable operator)
 * - memoize
 *
 * @ignore
 */
export function fastPropGetter(paths: string[]): (x: any) => any {
  const segments = paths;
  let seg = 'store.' + segments[0];
  let i = 0;
  const l = segments.length;

  let expr = seg;
  while (++i < l) {
    expr = expr + ' && ' + (seg = seg + '.' + segments[i]);
  }

  const fn = new Function('store', 'return ' + expr + ';');

  return <(x: any) => any>fn;
}

/**
 * Given an array of states, it will return a object graph. Example:
 *    const states = [
 *      Cart,
 *      CartSaved,
 *      CartSavedItems
 *    ]
 *
 * would return:
 *
 *  const graph = {
 *    cart: ['saved'],
 *    saved: ['items'],
 *    items: []
 *  };
 *
 * @ignore
 */
export function buildGraph(stateClasses: StateClass[]): StateKeyGraph {
  const findName = (stateClass: StateClass) => {
    const meta = stateClasses.find(g => g === stateClass);
    if (!meta) {
      throw new Error(`Child state not found: ${stateClass}`);
    }

    if (!meta[META_KEY]) {
      throw new Error('States must be decorated with @State() decorator');
    }

    return meta[META_KEY].name;
  };

  return stateClasses.reduce<StateKeyGraph>((result: StateKeyGraph, stateClass: StateClass) => {
    if (!stateClass[META_KEY]) {
      throw new Error('States must be decorated with @State() decorator');
    }

    const { name, children } = stateClass[META_KEY];
    result[name] = (children || []).map(findName);
    return result;
  }, {});
}

/**
 * Given a states array, returns object graph
 * returning the name and state metadata. Example:
 *
 *  const graph = {
 *    cart: { metadata }
 *  };
 *
 * @ignore
 */
export function nameToState(states: StateClass[]): ObjectKeyMap<StateClass> {
  return states.reduce<ObjectKeyMap<StateClass>>((result: ObjectKeyMap<StateClass>, stateClass: StateClass) => {
    if (!stateClass[META_KEY]) {
      throw new Error('States must be decorated with @State() decorator');
    }

    const meta = stateClass[META_KEY];
    result[meta.name] = stateClass;
    return result;
  }, {});
}

/**
 * Given a object relationship graph will return the full path
 * for the child items. Example:
 *
 *  const graph = {
 *    cart: ['saved'],
 *    saved: ['items'],
 *    items: []
 *  };
 *
 * would return:
 *
 *  const r = {
 *    cart: 'cart',
 *    saved: 'cart.saved',
 *    items: 'cart.saved.items'
 *  };
 *
 * @ignore
 */
export function findFullParentPath(obj: StateKeyGraph, newObj: ObjectKeyMap<string> = {}): ObjectKeyMap<string> {
  const visit = (child: StateKeyGraph, keyToFind: string): string => {
    for (const key in child) {
      if (child.hasOwnProperty(key) && child[key].indexOf(keyToFind) >= 0) {
        const parent = visit(child, key);
        return parent !== null ? `${parent}.${key}` : key;
      }
    }
    return null;
  };

  for (const key in obj) {
    if (obj.hasOwnProperty(key)) {
      const parent = visit(obj, key);
      newObj[key] = parent ? `${parent}.${key}` : key;
    }
  }

  return newObj;
}

/**
 * Given a object graph, it will return the items topologically sorted Example:
 *
 *  const graph = {
 *    cart: ['saved'],
 *    saved: ['items'],
 *    items: []
 *  };
 *
 * would return:
 *
 *  const results = [
 *    'items',
 *    'saved',
 *    'cart'
 *  ];
 *
 * @ignore
 */
export function topologicalSort(graph: StateKeyGraph): string[] {
  const sorted: string[] = [];
  const visited: ObjectKeyMap<boolean> = {};

  const visit = (name: string, ancestors: string[] = []) => {
    if (!Array.isArray(ancestors)) {
      ancestors = [];
    }

    ancestors.push(name);
    visited[name] = true;

    graph[name].forEach((dep: string) => {
      if (ancestors.indexOf(dep) >= 0) {
        throw new Error(`Circular dependency '${dep}' is required by '${name}': ${ancestors.join(' -> ')}`);
      }

      if (visited[dep]) {
        return;
      }

      visit(dep, ancestors.slice(0));
    });

    if (sorted.indexOf(name) < 0) {
      sorted.push(name);
    }
  };

  Object.keys(graph).forEach(k => visit(k));

  return sorted.reverse();
}

/**
 * Returns if the parameter is a object or not.
 *
 * @ignore
 */
export function isObject(obj) {
  return (typeof obj === 'object' && obj !== null) || typeof obj === 'function';
}
