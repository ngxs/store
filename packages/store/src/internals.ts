import { META_KEY } from './symbols';

export interface MetaDataModel {
  name: string;
  actions: { [key: string]: any[] };
  defaults: any;
  path: string;
  children: any[];
  instance: any;
}

/**
 * Ensures metadata is attached to the klass and returns it.
 */
export function ensureStoreMetadata(target): MetaDataModel {
  if (!target.hasOwnProperty(META_KEY)) {
    const defaultMetadata: MetaDataModel = {
      name: null,
      actions: {},
      defaults: {},
      path: null,
      children: [],
      instance: null
    };

    Object.defineProperty(target, META_KEY, { value: defaultMetadata });
  }

  return target[META_KEY];
}

/**
 * The generated function is faster than:
 * - pluck (Observable operator)
 * - memoize
 */
export function fastPropGetter(paths: string[]): (x: any) => any {
  const segments = paths;
  let seg = 'store.' + segments[0];
  let i = 0;
  const l = segments.length;

  let expr = seg;
  while (++i < l) {
    expr = expr + ' && ' + (seg = seg + '.' + segments[i]);
  }

  const fn = new Function('store', 'return ' + expr + ';');

  return <(x: any) => any>fn;
}

/**
 * Given an array of states, it will return a object graph. Example:
 *    const states = [
 *      Cart,
 *      CartSaved,
 *      CartSavedItems
 *    ]
 *
 * would return:
 *
 *  const graph = {
 *    cart: ['saved'],
 *    saved: ['items'],
 *    items: []
 *  };
 *
 */
export function buildGraph(states) {
  const findName = klass => {
    const meta = states.find(g => g === klass);
    if (!meta) {
      throw new Error(`Child state not found: ${klass}`);
    }

    if (!meta[META_KEY]) {
      throw new Error('States must be decorated with @State() decorator');
    }

    return meta[META_KEY].name;
  };

  return states.reduce((result, klass) => {
    if (!klass[META_KEY]) {
      throw new Error('States must be decorated with @State() decorator');
    }

    const { name, children } = klass[META_KEY];
    result[name] = (children || []).map(findName);
    return result;
  }, {});
}

/**
 * Given a states array, returns object graph
 * returning the name and state metadata. Example:
 *
 *  const graph = {
 *    cart: { metadata }
 *  };
 *
 */
export function nameToState(states) {
  return states.reduce((result, klass) => {
    if (!klass[META_KEY]) {
      throw new Error('States must be decorated with @State() decorator');
    }

    const meta = klass[META_KEY];
    result[meta.name] = klass;
    return result;
  }, {});
}

/**
 * Given a object relationship graph will return the full path
 * for the child items. Example:
 *
 *  const graph = {
 *    cart: ['saved'],
 *    saved: ['items'],
 *    items: []
 *  };
 *
 * would return:
 *
 *  const r = {
 *    cart: 'cart',
 *    saved: 'cart.saved',
 *    items: 'cart.saved.items'
 *  };
 *
 */
export function findFullParentPath(obj: any, newObj: any = {}) {
  const visit = (child: any, keyToFind: string) => {
    for (const key in child) {
      if (child.hasOwnProperty(key) && child[key].indexOf(keyToFind) >= 0) {
        const parent = visit(child, key);
        return parent !== null ? `${parent}.${key}` : key;
      }
    }
    return null;
  };

  for (const key in obj) {
    if (obj.hasOwnProperty(key)) {
      const parent = visit(obj, key);
      newObj[key] = parent ? `${parent}.${key}` : key;
    }
  }

  return newObj;
}

/**
 * Given a object graph, it will return the items topologically sorted Example:
 *
 *  const graph = {
 *    cart: ['saved'],
 *    saved: ['items'],
 *    items: []
 *  };
 *
 * would return:
 *
 *  const results = [
 *    'items',
 *    'saved',
 *    'cart'
 *  ];
 *
 */
export function topologicalSort(graph) {
  const sorted = [];
  const visited = {};
  const visit = (name, ancestors: any = []) => {
    if (!Array.isArray(ancestors)) {
      ancestors = [];
    }

    ancestors.push(name);
    visited[name] = true;

    graph[name].forEach(dep => {
      if (ancestors.indexOf(dep) >= 0) {
        throw new Error(`Circular dependency '${dep}' is required by '${name}': ${ancestors.join(' -> ')}`);
      }

      if (visited[dep]) {
        return;
      }

      visit(dep, ancestors.slice(0));
    });

    if (sorted.indexOf(name) < 0) {
      sorted.push(name);
    }
  };

  Object.keys(graph).forEach(k => visit(k));

  return sorted.reverse();
}

/**
 * Returns if the parameter is a object or not.
 */
export function isObject(obj) {
  return (typeof obj === 'object' && obj !== null) || typeof obj === 'function';
}
