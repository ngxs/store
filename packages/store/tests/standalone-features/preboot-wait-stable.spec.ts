import { JsonPipe } from '@angular/common';
import {
  AfterViewInit,
  Component,
  Injectable,
  inject,
  ɵPendingTasks as PendingTasks,
  ɵprovideZonelessChangeDetection,
  ChangeDetectionStrategy
} from '@angular/core';
import { bootstrapApplication } from '@angular/platform-browser';
import { renderApplication } from '@angular/platform-server';
import {
  Action,
  ActionStatus,
  Actions,
  State,
  StateContext,
  StateToken,
  dispatch,
  provideStore,
  select,
  withNgxsPreboot
} from '@ngxs/store';
import { freshPlatform } from '@ngxs/store/internals/testing';

function executeRecipeFromDocs() {
  const pendingTasks = inject(PendingTasks);
  const actions$ = inject(Actions);

  const actionToTaskIdMap = new Map<any, number>();

  actions$.subscribe(ctx => {
    if (ctx.status === ActionStatus.Dispatched) {
      const taskId = pendingTasks.add();
      actionToTaskIdMap.set(ctx.action, taskId);
    } else {
      const taskId = actionToTaskIdMap.get(ctx.action);
      if (typeof taskId === 'number') {
        pendingTasks.remove(taskId);
        actionToTaskIdMap.delete(ctx.action);
      }
    }
  });
}

describe('preboot feature + stable', () => {
  const COUNTRIES_STATE_TOKEN = new StateToken<string[]>('countries');

  class AddCountry {
    static type = '[Countries] Add country';
    constructor(public country: string) {}
  }

  @State({
    name: COUNTRIES_STATE_TOKEN,
    defaults: ['Angola', 'Namibia', 'Botswana']
  })
  @Injectable()
  class CountriesState {
    @Action(AddCountry)
    async addCountry(ctx: StateContext<string[]>, action: AddCountry) {
      await new Promise(resolve => setTimeout(resolve, 0));
      ctx.setState(state => [...state, action.country]);
    }
  }

  @Component({
    selector: 'app-root',
    template: `<div>{{ countries() | json }}</div>`,
    changeDetection: ChangeDetectionStrategy.OnPush,
    standalone: true,
    imports: [JsonPipe]
  })
  class TestComponent implements AfterViewInit {
    countries = select(COUNTRIES_STATE_TOKEN);

    private addCountry = dispatch(AddCountry);

    ngAfterViewInit(): void {
      this.addCountry('Zambia');
    }
  }

  it(
    'should wait for app to become stable',
    freshPlatform(async () => {
      // Arrange
      const html = await renderApplication(
        () =>
          bootstrapApplication(TestComponent, {
            providers: [
              ɵprovideZonelessChangeDetection(),

              provideStore([CountriesState], withNgxsPreboot(executeRecipeFromDocs))
            ]
          }),
        {
          document: '<app-root></app-root>',
          url: '/'
        }
      );

      // Assert
      expect(html).toMatchSnapshot();
    })
  );
});
