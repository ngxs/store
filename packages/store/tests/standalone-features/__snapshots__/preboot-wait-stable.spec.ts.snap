// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preboot feature + stable should wait for app to become stable 1`] = `
"<html><head></head><body><app-root ng-version="17.3.3" ng-server-context="other"><div>[
  "Angola",
  "Namibia",
  "Botswana",
  "Zambia"
]</div></app-root></body></html>"
`;
